pkg/vuln/nvdapi/detector.go

```diff
@@ -0,0 +1,434 @@
+
+				vr, err := convertCpeMatchToVersionRange(cpeMatch)
+				if err != nil {
+					return nil, err
+				}
+
+				m := vuln.Match{
+					Package: vuln.Package{
+						Name: packageName,
+					},
+					CPE: vuln.CPE{
+						URI:          cpeMatch.Criteria,
+						VersionRange: vr,
+					},
+					Vulnerability: vuln.Vulnerability{
+						ID:  cve.ID,
+						URL: fmt.Sprintf("https://nvd.nist.gov/vuln/detail/%s", cve.ID),
+					},
+				}
+
+				return &m, nil
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+func cpeStringsMatch(requestCPE, responseCPE string) (bool, error) {
+	req, err := wfn.Parse(requestCPE)
+	if err != nil {
+		return false, fmt.Errorf("cannot compare CPEs: %w", err)
+	}
+	resp, err := wfn.Parse(responseCPE)
+	if err != nil {
+		return false, fmt.Errorf("cannot compare CPEs: %w", err)
+	}
+
+	// Return "false" when the request CPE had no TargetSW set, but the response CPE
+	// did have one set. This case usually means that the response CVE is for a
+	// plugin or language-specific package. (For example, a Jenkins plugin.)
+	if req.TargetSW == wfn.Any && resp.TargetSW != wfn.Any {
+		return false, nil
+	}
+
+	return wfn.Match(req, resp), nil
+}
+
+var ErrRateLimited = errors.New("we've been rate limited by NVD! 🙊")
+
+func (s *Detector) doSearch(ctx context.Context, cpe string) ([]Cve, error) {
+	err := s.rateLimiter.Wait(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Deal with pages (not urgent because the default page size is 2,000
+	//  CVEs, and we're searching for single packages at a time.)
+
+	// TODO: Consider using an in-memory cache of API responses, keyed by the
+	//  requested CPE. This would save time on redundant API requests, like searching
+	//  for '...*:go...' multiple times because we've pruned versions from multiple,
+	//  related packages like 'go-1.18', 'go-1.19', and 'go-1.20'.
+
+	reqURL := fmt.Sprintf(
+		"https://%s%s?virtualMatchString=%s",
+		s.serviceHost,
+		s.serviceEndpoint,
+		cpe,
+	)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create request with URL %q: %w", reqURL, err)
+	}
+
+	req.Header["Accept"] = []string{"application/json"}
+	req.Header["User-Agent"] = []string{"wolfictl"}
+
+	if k := s.apiKey; k != "" {
+		req.Header["apiKey"] = []string{s.apiKey}
+	}
+
+	log.Printf("☎️  sending API request: %s", reqURL)
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("unable to complete request to URL %q: %w", reqURL, err)
+	}
+	if s := resp.StatusCode; s != http.StatusOK {
+		if s == http.StatusForbidden || s == http.StatusTooManyRequests {
+			return nil, ErrRateLimited
+		}
+
+		return nil, fmt.Errorf("got unexpected response status %d for request to %q. Headers: %+v", s, reqURL, resp.Header)
+	}
+
+	defer resp.Body.Close()
+
+	dec := json.NewDecoder(resp.Body)
+	var cvesResponse CVEsResponse
+	err = dec.Decode(&cvesResponse)
+	if err != nil {
```

Comment:
```suggestion
	var cvesResponse CVEsResponse
	if err := json.NewDecoder(resp.Body).Decode(&cvesResponse); err != nil {
```

(super nit, but it's a lot less code and I think it scans better)

Reply:
❤️